kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt

```diff
@@ -0,0 +1,217 @@
+          removeAllViews()
+          addView(it.viewForMain(screens, viewRegistry, this))
+        }
+    )
+
+    // Compare the new modals list to the set of dialogs already on display. Show
+    // any new ones, throw away any stale ones.
+    takeScreensSubs.add(screens
+        .whileAttached()
+        .subscribe { mainAndModal ->
+          val newDialogs = mutableListOf<DialogRef>()
+          for ((i, modalScreen) in mainAndModal.modals.withIndex()) {
+            newDialogs +=
+              if (dialogs.size < i && dialogs[i].screen::class == modalScreen::class) {
+                dialogs[i]
+              } else {
+                DialogRef(modalScreen, mainAndModal.showDialog(i, screens, viewRegistry, this))
+              }
+          }
+
+          dialogs.filter { it !in newDialogs }
+              .forEach { it.dialog.hide() }
+          dialogs = newDialogs
+        }
+    )
+  }
+
+  private fun showAlert(alertScreen: AlertScreen): AlertDialog {
+    val builder = AlertDialog.Builder(context)
+
+    if (alertScreen.cancelable) {
+      builder.setOnCancelListener { alertScreen.onEvent(Canceled) }
+    } else {
+      builder.setCancelable(false)
+    }
+
+    for ((button, name) in alertScreen.buttons) {
+      when (button) {
+        POSITIVE -> builder.setPositiveButton(name) { _, _ ->
+          alertScreen.onEvent(ButtonClicked(POSITIVE))
+        }
+        NEGATIVE -> builder.setNegativeButton(name) { _, _ ->
+          alertScreen.onEvent(ButtonClicked(NEGATIVE))
+        }
+        NEUTRAL -> builder.setNeutralButton(name) { _, _ ->
+          alertScreen.onEvent(ButtonClicked(NEUTRAL))
+        }
+      }
+    }
+
+    alertScreen.message.takeIf { it.isNotEmpty() }
+        .let { builder.setMessage(it) }
+    alertScreen.title.takeIf { it.isNotEmpty() }
+        .let { builder.setTitle(it) }
+
+    return builder.show()
+  }
+
+  private fun <T> Observable<T>.whileAttached(): Observable<T> =
+    attached.switchMap { isAttached -> if (isAttached) this else never<T>() }
+
+  private class DialogRef(
+    val screen: Any,
+    val dialog: Dialog
+  )
+
+  private fun <D : Any> MainAndModalScreen<*, D>.showDialog(
+    index: Int,
+    mainAndModalScreens: Observable<out MainAndModalScreen<*, *>>,
+    viewRegistry: ViewRegistry,
+    container: ViewGroup
+  ): Dialog {
+    (modals[index] as? AlertScreen)?.let { return showAlert(it) }
+
+    // https://github.com/square/workflow/issues/58
+    TODO("Non-alerts are not supported yet.")
+//    val modalScreens: Observable<out D> = mainAndModalScreens.mapToModalIndexMatching(index, this)
+//    val binding:
+  }
+
+  companion object : ViewBinding<MainAndModalScreen<*, *>>
+  by BuilderBinding(
+      type = MainAndModalScreen::class.java,
+      builder = { screens, builders, context, _ ->
+        MainAndModalContainer(context).apply {
+          layoutParams = (ViewGroup.LayoutParams(
+              android.view.ViewGroup.LayoutParams.MATCH_PARENT,
```

Comment:
Nit: Can we static-import these?

Reply:
I dropped the `android.view`, but left in `ViewGroup.LayoutParams`. IIRC you need to be _very_ careful to get the right type w/these things.

Reply:
NM, that's covered by the constructor. Imported `MATCH_PARENT`.